satellite/repair/checker/checker.go

```diff
@@ -221,18 +182,97 @@ func (checker *Checker) updateSegmentStatus(ctx context.Context, pointer *pb.Poi
 		// we need one additional piece for error correction. If only the minimum is remaining the file can't be repaired and is lost.
 		// except for the case when minimum and repair thresholds are the same (a case usually seen during testing)
 	} else if numHealthy <= redundancy.MinReq && numHealthy < redundancy.RepairThreshold {
+
+		// make an entry in to the irreparable table
+		segmentInfo := &pb.IrreparableSegment{
+			Path:               []byte(path),
+			SegmentDetail:      pointer,
+			LostPieces:         int32(len(missingPieces)),
+			LastRepairAttempt:  time.Now().Unix(),
+			RepairAttemptCount: int64(1),
+		}
+
+		// add the entry if new or update attempt count if already exists
+		err := checker.irrdb.IncrementRepairAttempts(ctx, segmentInfo)
+		if err != nil {
+			return Error.New("error handling irreparable segment to queue %s", err)
+		}
+	}
+	return nil
+}
+
+// checkerObserver implements the metainfo loop Observer interface
+type checkerObserver struct {
+	repairQueue queue.RepairQueue
+	irrdb       irreparable.DB
+	nodestate   *ReliabilityCache
+	monStats    durabilityStats
+	log         *zap.Logger
+}
+
+func (obs *checkerObserver) RemoteSegment(ctx context.Context, path storj.Path, pointer *pb.Pointer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	obs.monStats.remoteSegmentsChecked++
+	remote := pointer.GetRemote()
+
+	pieces := remote.GetRemotePieces()
+	if pieces == nil {
+		obs.log.Debug("no pieces on remote segment")
+		return nil
+	}
+
+	missingPieces, err := obs.nodestate.MissingPieces(ctx, pointer.CreationDate, pieces)
+	if err != nil {
+		return Error.New("error getting missing pieces %s", err)
+	}
+
+	numHealthy := int32(len(pieces) - len(missingPieces))
+	redundancy := pointer.Remote.Redundancy
+	mon.IntVal("checker_segment_total_count").Observe(int64(len(pieces)))
+	mon.IntVal("checker_segment_healthy_count").Observe(int64(numHealthy))
+
+	// we repair when the number of healthy pieces is less than or equal to the repair threshold
+	// except for the case when the repair and success thresholds are the same (a case usually seen during testing)
+	if numHealthy > redundancy.MinReq && numHealthy <= redundancy.RepairThreshold && numHealthy < redundancy.SuccessThreshold {
+		if len(missingPieces) == 0 {
+			obs.log.Warn("Missing pieces is zero in checker, but this should be impossible -- bad redundancy scheme.")
+			return nil
+		}
+		obs.monStats.remoteSegmentsNeedingRepair++
+		err = obs.repairQueue.Insert(ctx, &pb.InjuredSegment{
+			Path:         []byte(path),
+			LostPieces:   missingPieces,
+			InsertedTime: time.Now().UTC(),
+		})
+		if err != nil {
+			obs.log.Sugar().Errorf("error adding injured segment to queue %s", err)
+			return nil
+		}
+
+		// delete always returns nil when something was deleted and also when element didn't exists
+		err = obs.irrdb.Delete(ctx, []byte(path))
+		if err != nil {
+			obs.log.Sugar().Errorf("error deleting entry from irreparable db: ", zap.Error(err))
```

Comment:
There is no need to use `Sugar` here, you can do the same with `obs.log.Error` and it's mostly what we want.